Add proper versioning for both the raw and the preprocessed profile format

mstange opened the pull request:
Fixes #165.
This setup allows us to make changes to either profile format and still keep
compatibility with old profiles.

When the raw profile format changes (due to a change in Gecko):
 - The version number (profile.meta.version) needs to be incremented in Gecko.
 - In raw-profile-versioning.js, CURRENT_VERSION needs to be set to that
   incremented version, and a conversion function from the old to the new
   version needs to be added to _updaters.
 - Profile preprocessing may need to be adjusted to parse the new raw
   profile format version.

When the preprocessed profile format changes (e.g. because a different data
format seems adequate, or because new data from the raw profile needs to be
accommodated):
 - preprocessProfile needs to be changed to output the new format.
 - In preprocessed-profile-versioning.js, CURRENT_VERSION needs to be
   incremented and an update function needs to be added to _updaters.

At all times, preprocessProfile only has code that converts the latest version
of the raw profile format into the latest version of the preprocessed profile
format. All compatibility code is hosted outside of preprocess-profile.js, in
the respective *-profile-versioning.js file.

gregtatum commented on src/content/data-table-utils.js:
```diff
@@ -0,0 +1,66 @@
+ *                               Array.prototype.sort.
+ * @returns The data table.
+ */
+export function sortDataTable(table, keyColumn, comparator) {
```
This makes my palms sweat and makes me think of coding interviews. 😂

mstange replied:
Haha, I just used Google...

gregtatum commented on src/content/old-cleopatra-profile-format.js:
```diff
@@ -0,0 +1,299 @@
+ * We want preprocess-profile to be exclusively concerned with converting the
+ * most recent raw profile format into the most recent preprocessed profile
+ * format.
+ * Some of the code below is basically a snapshot of the preprocessing code as
```
👍 nice approach.

gregtatum commented on src/content/data-table-utils.js:
```diff
@@ -0,0 +1,66 @@
+/**
```
Can we flow type new files? That way we won't have as much dedicated flow work.

gregtatum commented on src/content/preprocessed-profile-versioning.js:
```diff
@@ -0,0 +1,89 @@
+
+  if (profileVersion > CURRENT_VERSION) {
+    throw new Error(
+      `Unable to parse a preprocessed profile of version ${profileVersion} - are you running an outdated version of perf.html? ` +
```
Thanks for taking this case into account.

gregtatum commented on src/content/preprocessed-profile-versioning.js:
```diff
@@ -0,0 +1,89 @@
+// _upgraders[i] converts from version i - 1 to version i.
+// Every "upgrader" takes the profile as its single argument and mutates it.
+const _upgraders = {
+  1: profile => {
```
This is valid syntax?! Well... today I learned something.

mstange replied:
It's a bit of a lie, since object property names are implicitly stringified anyway... and now that I'm flow-checking this file, it's complaining about it, so I'll make them string keys.

mstange replied:
Ok, I'm going to use the [1]: ... syntax.

gregtatum commented on src/content/preprocessed-profile-versioning.js:
```diff
@@ -0,0 +1,89 @@
+      sortDataTable(thread.markers, thread.markers.time, (a, b) => a - b);
+    }
+
+    // And threads have proper names and processType fields.
```
Annotating with comments will be really nice explaining the history of our profile format changes.

gregtatum commented on src/content/preprocessed-profile-versioning.js:
```diff
@@ -0,0 +1,89 @@
+/**
```
Another new file, good candidate for flow typing.

mstange commented:
I've updated the pull request. I've written some documentation that I'm actually happy with, and I've added flow to the new files. Thanks for pushing for that, it really improves things.

mstange commented:
I also wrote a test for `sortDataTable`, and obviously it found a bug...

gregtatum commented on docs/upgrading-profiles.md:
```diff
@@ -0,0 +1,80 @@
+
+We'll see how this works out. It might be annoying for users who have updated their Firefox Nightly but not opened up perf.html between (2) and (3), and then try to load a new profile into an old version of perf.html from their service worker cache. They're just one page refresh away from success, though.
+
+If this turns out to be too hard to deal with, we might need to go with a semver approach and hope that most profile changes are non-breaking. Time will tell.
```
Yeah, I was considering semver too, but for now I like the single number bump.

gregtatum commented on docs/upgrading-profiles.md:
```diff
@@ -0,0 +1,80 @@
+
+### The old cleopatra format
+
+The "old cleopatra format" is the profile format that was used by the cleopatra version before the big rewrite. Profiles of this format are still in the profile store, and there are links to those profiles strewn accross bugzilla. We want to be able to display those profiles.
```
s/accross/across/
